File: 3D/src/custom_models/KeyTransformation.java
package custom_models;

import org.joml.Quaternionf;
import org.joml.Vector2i;
import org.joml.Vector3f;
import org.joml.Vector3i;

import kmerrill285.Inignoto.resources.raytracer.Raytracer;

public class KeyTransformation {
	public static final float SCALING = 1.0f / 32.0f;
	
	public Vector3f position = new Vector3f(0, 0, 0);
	public Quaternionf rotation = new Quaternionf().identity();
	public Vector3f scale = new Vector3f(0, 0, 0);
	public Vector3i size = new Vector3i(0, 0, 0);
	public Vector2i uv = new Vector2i(0, 0);
	public Vector3f origin = new Vector3f(0, 0, 0);
		
	public boolean visible = true;
	public boolean locked = true;
		
	public Vector3f look = new Vector3f(0, 0, 1);
	
	public Vector3f axisAngles = new Vector3f(0, 0, 0);
	
	
	public void translate(Vector3f translation) {
		this.position.add(translation);
	}
	
	public void rotate(Vector3f rotation) {
		rotate(rotation, new Vector3f(this.position).add(new Vector3f(this.origin).rotate(this.rotation)));
	}
	
	public void rotate(Vector3f rotation, Vector3f origin) {
		
		this.axisAngles.add(rotation);
		
        this.rotation.rotateLocalX((float)Math.toRadians(rotation.x));
        this.rotation.rotateLocalY((float)Math.toRadians(rotation.y));
        this.rotation.rotateLocalZ((float)Math.toRadians(rotation.z));
                
        Quaternionf nRot = new Quaternionf().identity();
        nRot.rotateLocalX((float)Math.toRadians(rotation.x));
        nRot.rotateLocalY((float)Math.toRadians(rotation.y));
        nRot.rotateLocalZ((float)Math.toRadians(rotation.z));
        
       
        this.position = Raytracer.rotateAround(this.position, origin, nRot);
        
    }
	
	public void setRotation(Vector3f rotation) {
		rotate(new Vector3f(rotation).mul(-1));
		this.axisAngles = new Vector3f(0, 0, 0);
		this.rotation.identity();
		rotate(rotation);
	}
	
	public KeyTransformation copy() {
		KeyTransformation p = new KeyTransformation();
		p.position = new Vector3f(position);
		p.rotation = new Quaternionf(rotation);
		p.scale = new Vector3f(scale);
		p.size = new Vector3i(size);
		p.uv = new Vector2i(uv);
		p.origin = new Vector3f(origin);
		return p;
	}
}

File: MonoGame/Inignoto/Graphics/Models/KeyTransformation.cs
using Inignoto.Math;
using Microsoft.Xna.Framework;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Inignoto.Graphics.Models
{
    public class KeyTransformation
    {
        public Vector3 position = new Vector3();
        public Quaternion rotation = new Quaternion();
        public Vector3 scale = new Vector3(1, 1, 1);
        public Vector3 size = new Vector3();
        public Vector2 uv = new Vector2(0, 0);
        public Vector3 origin = new Vector3();

        public bool visible = true;
        public bool locked = true;

        public Vector3 look = new Vector3(0, 0, 1);
        public Vector3 axisAngles = new Vector3(0, 0, 0);

        public void Translate(Vector3 translation)
        {
            position += translation;
        }

        public void Rotate(Vector3 rotation)
        {
            Rotate(Quaternion.CreateFromYawPitchRoll(rotation.Y, rotation.X, rotation.Z), position + (Matrix.CreateTranslation(origin) * Matrix.CreateFromQuaternion(this.rotation)).Translation);
        }

        public void Rotate(Quaternion rot, Vector3 origin)
        {
            rotation = Quaternion.Concatenate(rotation, rot);
            
            position = Raytracing.RotateAround(position, origin, new Quaternion(rot.X, rot.Y, rot.Z, rot.W));

        }

        public void Rotate(Vector3 rotation, Vector3 origin)
        {
            Quaternion xRot = Quaternion.CreateFromAxisAngle(new Vector3(1, 0, 0), rotation.X);
            Quaternion yRot = Quaternion.CreateFromAxisAngle(new Vector3(0, 1, 0), rotation.Y);
            Quaternion zRot = Quaternion.CreateFromAxisAngle(new Vector3(0, 0, 1), rotation.Z);

            Quaternion rot = Quaternion.Concatenate(Quaternion.Concatenate(xRot, yRot), zRot);

            this.rotation = Quaternion.Concatenate(this.rotation, rot);
            axisAngles += rotation;

            position = Raytracing.RotateAround(position, origin, new Quaternion(rot.X, rot.Y, rot.Z, rot.W));

        }

        public void RotateEuler(Vector3 rotation, Vector3 origin)
        {
            Quaternion rot = Quaternion.CreateFromYawPitchRoll(rotation.Y, rotation.X, rotation.Z);

            this.rotation = Quaternion.CreateFromYawPitchRoll(rotation.Y, rotation.X, rotation.Z);

            position = Raytracing.RotateAround(position, origin, new Quaternion(rot.X, rot.Y, rot.Z, rot.W));

        }

        public void SetRotation(Vector3 rotation)
        {
            Rotate(new Vector3(rotation.X, rotation.Y, rotation.Z) * -1);
            axisAngles *= 0;
            this.rotation = Quaternion.Identity;
            Rotate(rotation);
        }

        public KeyTransformation Copy()
        {
            KeyTransformation p = new KeyTransformation
            {
                position = new Vector3(position.X, position.Y, position.Z),
                rotation = new Quaternion(rotation.X, rotation.Y, rotation.Z, rotation.W),
                scale = new Vector3(scale.X, scale.Y, scale.Z),
                size = new Vector3(size.X, size.Y, size.Z),
                uv = new Vector2(uv.X, uv.Y),
                origin = new Vector3(origin.X, origin.Y, origin.Z)
            };
            return p;
        }
    }
}
